### ISSUE TITLE
[XrdSecgsi] regression in 47eb688dc2131c30edd57eba525f95bb2ea842ba

### ISSUE THREAD

**Issue by simonmichal** [xrootd/xrootd#725]
While testing the recent changes in XrdSecgsi I noticed that a regression has been first introduced in 47eb688dc2131c30edd57eba525f95bb2ea842ba (and **haven't been fixed** by 1533382e11418c24916c93392973ac1c3fcab552).

Consider XRootD with CERN Grid host certificate running on host.cern.ch. When the client tries to authenticate after issuing a command like:
```
 xrdfs host ls /tmp
```
(note hostname was given without _.cern.ch_ domain)
it fails and following error can be seen in the logs:

```
secgsi_getCredentials: server certificate CN 'host.cern.ch' does not match the expected format(s): '[*/]host[/*]' (default); exceptions are controlled by the env XrdSecGSISRVNAMES CF: 0x7fffeee1c360
secgsi_ErrF: Secgsi: ErrParseBuffer: server certificate CN 'host.cern.ch' does not match the expected format(s): '[*/]host[/*]' (default); exceptions are controlled by the env XrdSecGSISRVNAMES: kXGS_cert
```

@bbockelm : Before the commit in question the _host_ hostname was properly resolved into _host.cern.ch_ , currently we do DNS lookups only for IP addresses. Could you explain why the lookup has been removed for plain hostnames?

**Comment by xrootd-dev** [comment]
I think the issue here is a misunderstanding on what isHostName() does. It 
simply veriies that the string you pass is not an RFC conforming IP 
address. It does nothing else. It's used simply to decide twhch context to 
apply to the string. In gsi case, if isHostName() returns true, you still 
have to use a reverse lookup to valdate that the presumed hostname is 
correct and to fully qualify it. That's what the original code did but 
without the test and it always done presuming that if you didn't want to 
use the hostname then you wouldn't. It wasn't the most effecient solution 
but the again it wasn't too much added overhead to gsi anyway.

BTW the reason we introduced the reverse lookup was because if you request 
nodnr (don't do reverse lookups -- which most sites specify as it's a 
no-brainer to speed things up, then hostname would always be the IP 
address which wasn't particularly useful for gsi. Having SAN makes it a 
bit more complicated.

Andy

**Comment by bbockelm** [comment]
Hi,

I think the interpretation of `isHostName` is correct - it's meant to detect and handle the case where the user provides a raw IP address.  In such a case, we rely on reverse DNS.

If we want to copy what the rest of the world does, we would fail the particular use case mentioned by @simonmichal .  For example, `curl` does _not_ expand a hostname provided by the user to a FQDN:

```
$ curl https://www/
curl: (51) Unable to communicate securely with peer: requested domain name does not match the server's certificate.
```

This is because DNS is typically considered insecure and `curl` doesn't want to rely on DNS to bootstrap the TLS security.  On this host, the DNS search domain is set to `unl.edu`, meaning that `www` expands to `www.unl.edu`; the latter matches the server certificate.

Ways forward:
1.  At the GSI layer, additionally detect non-qualified hostnames and qualify them.
   * From what I can tell, this would continue to make GSI "unique" -- KRB5 doesn't do this currently.
   * This means GSI is reliant on DNS still, which is bad from the security standpoint.  But this occurs anyway in order to not break EOS support.
2.  At the client layer (`xrdfs`, `xrdcp`) detect non-qualified hostnames and qualify them.
3.  Have a backward compat break.

I'd lean toward (1) (maybe adding a security warning or a config flag to allow folks to turn off the reliance on DNS?).

Thoughts?

Brian

PS - @simonmichal - you may want to let the CMS EOS be aware of the fact that they don't have SANs setup correctly.  That's very fixable but would cause problems the improved security.

**Comment by gganis** [comment]
Hi Brian,
I also lean for (1), more centralized and simplifying deployment for the "fix".
I support the idea of a config flag to allow to turn off the DNS lookup.

Gerri

**Comment by bbockelm** [comment]
Option (1) is implemented in #731.  @simonmichal, please test this with your setup.
